### PR TITLE
Resource Removal: Add resource Group support for mg level deployments

### LINKED ISSUE #1042: Resource Removal issue when the root module (deploy.bicep) is management group scope and child module (deploy.bicep) is resource group scope
For modules that support (resource group scope) in the 'Microsoft.Authorization' namespace. Attempting to deploy the pipelines when 'publishing is set to true', will break for the parameter files that target 'resource group' scope deployments.

### Scenario

When having parameter files in the 'Microsoft.Authorization' namespace modules, which perform deployments at resource group scope: 
- policyAssignments
- policyExemptions
- roleDefinitions
- roleAssignments

If I set modify the parameter file to target resource group like this:
```json
{
    "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#",
    "contentVersion": "1.0.0.0",
    "parameters": {
        "roleName": {
            "value": "<<namePrefix>>-az-testRole-rg-min"
        },
        "actions": {
            "value": [
                "Microsoft.Compute/galleries/read",
                "Microsoft.Compute/galleries/images/read"
            ]
        },
        "subscriptionId": {
            "value": "<<subscriptionId>>"
        },
        "resourceGroupName": {
            "value": "<<resourceGroupName>>"
        }
    }
}
```

The root deploy.bicep, will call the nested module that performs deployment at the 'resource group scope'. The template deployment is successful. However the 'removal step' does not function for this type of deployment. Note that management group and subscription scopes are ok.

This is what happens inside the log for the deployment validation:
```pwsh
VERBOSE: Did not to find deployments by name [policyAssignments-20220223T0402139248Z] in scope [managementgroup]. Retrying in [60] seconds [6/40]
```
image

### Steps to reproduce

1. Create a parameter file that scopes to the resource group, similar to the one above. 
2. Update the workflow for that module to include the parameter file
3. Run the workflow on the branch by keeping 'remove deployed module' to `true`

### Workaround

Change the 'remove deployed module' to have default set to `false` for these modules so that when a merge to main happens, the deployment action is successful and publishing works

### PR DESCRIPTION
**Pull request by ahmadabdalla**
Closes #1042 

# Change

- Allowed 'Resource Group Removal' within a 'Management Group' Scope deployment.

Without this, removal stage for deployments that happen at the management group scope and target resource group as child will not be able to find the right deployment as the original one only supports subscription as a child 

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

**Comment by ahmadabdalla** [comment]
## Log Evidence
```powershell
VERBOSE: Invoke task with
  VERBOSE: {
    "ManagementGroupId": "***",
    "DeploymentName": "roleDefinitions-20220225T0502569172Z",
    "ResourceGroupName": "validation-rg",
    "Verbose": true,
    "TemplateFilePath": "/home/runner/work/ResourceModules/ResourceModules/arm/Microsoft.Authorization/roleDefinitions/deploy.bicep"
  }
  
  VERBOSE: Template file path: [/home/runner/work/ResourceModules/ResourceModules/arm/Microsoft.Authorization/roleDefinitions/deploy.bicep]
  VERBOSE: Module name: [roleDefinitions]
  VERBOSE: Handling resource removal with deployment name [roleDefinitions-20220225T0502569172Z]
  VERBOSE: Determined deployment scope [managementgroup]
  VERBOSE: Found deployed resource [/subscriptions/***/resourceGroups/validation-rg/providers/Microsoft.Authorization/roleDefinitions/2abcf71f-a840-5ac1-bac2-9d12[546](https://github.com/Azure/ResourceModules/runs/5329519376?check_suite_focus=true#step:4:546)e0ba9]
  VERBOSE: Total number of deployment target resources after fetching deployments [1]
  VERBOSE: Total number of deployment target resources  after pre-filtering (duplicates) & ordering items [1]
  VERBOSE: Total number of deployment target resources after formatting items [1]
  VERBOSE: Using local tokens [namePrefix]
  VERBOSE: Total number of deployments after filtering all dependency resources [1]
  VERBOSE: Total number of deployments after final ordering of resources [1]
  VERBOSE: Performing the operation "Remove" on target "[1] resources".
  VERBOSE: Performing the operation "Remove" on target "[1] Resource(s) with a maximum of [3] attempts.".
  VERBOSE: - Remove [/subscriptions/***/resourceGroups/validation-rg/providers/Microsoft.Authorization/roleDefinitions/2abcf71f-a840-5ac1-bac2-9d12546e0ba9]
  VERBOSE: ----------------------------------
  VERBOSE: Removing resource [2abcf71f-a840-5ac1-bac2-9d12546e0ba9] of type [Microsoft.Authorization/roleDefinitions]
  VERBOSE: Performing the operation "Remove" on target "Resource [/subscriptions/***/resourceGroups/validation-rg/providers/Microsoft.Authorization/roleDefinitions/2abcf71f-a840-5ac1-bac2-9d12546e0ba9]".
  VERBOSE: Resource ID [/subscriptions/***/resourceGroups/validation-rg/providers/Microsoft.Authorization/roleDefinitions/2abcf71f-a840-5ac1-bac2-9d12546e0ba9]
  VERBOSE: Resource Type [Microsoft.Authorization/roleDefinitions]
  VERBOSE: Performing the operation "Execute" on target "Post-resource-removal for [/subscriptions/***/resourceGroups/validation-rg/providers/Microsoft.Authorization/roleDefinitions/2abcf71f-a840-5ac1-bac2-9d12546e0ba9]".
  VERBOSE: ----------------------------------
  VERBOSE: The removal completed successfully
```

## Pipeline tested

Tested on the Role Definition Pipeline by adding a new parameter that targets Resource Group. Switched the workflow to use the root deploy.bicep and then reverted the configuration

File: utilities/pipelines/resourceRemoval/helper/Get-DeploymentTargetResourceList.ps1 [hunk 1]
```diff
@@ -108,7 +108,14 @@ function Get-DeploymentTargetResourceListInner {
                 [array]$resultSet = $resultSet | Where-Object { $_ -ne $deployment }
                 if ($deployment -match '/subscriptions/') {
                     # Subscription Level Child Deployments
-                    [array]$resultSet += Get-DeploymentTargetResourceListInner -Name (Split-Path $deployment -Leaf) -Scope 'subscription'
+                    if ($deployment -match '/resourceGroups/') {
```
Comment (MrMCake) [inline utilities/pipelines/resourceRemoval/helper/Get-DeploymentTargetResourceList.ps1:111]: I'd like to understand this better. Wouldn't the passed in ID be first matching `/subscription/` and hence be passed into the recrusive function with scope `subscription` anyways where it then checks for `if ($deployment -match '/resourceGroups/')` the same way we do? 

I assume somewhere down the line the deploymentId is not properly resolved, but I wonder why.

Can you e.g. provide an example for a deployment ID which would not be properly passed down?

Comment (ahmadabdalla) [reply utilities/pipelines/resourceRemoval/helper/Get-DeploymentTargetResourceList.ps1:111]: Here is an example:

`/subscriptions/20d6fbfe-b049-471c-95af-1369d14d0d45/resourceGroups/validation-rg/providers/Microsoft.Resources/deployments/l4aytak3mfire-PolicyAssignment-RG-Module`

Flow:
1- scope = managementGroup
2- deploymentTarget = GetAzManagementGroupdeploymentOperation --> for deployment Name.. SUCCESS and returns the above ID
3- foreach deployment in that deploymentTarget
4- does it match /subscriptions/ --> success
5- Reruns the TargetResourceListInner at the subscriptionScope
6- deploymentTarget = Get-AzDeploymentOperation --> for deployment Name.. FAIL --> That does not exist at this level (only on Mgmt Group)

Hence it does not even reach the resource group level

Comment (MrMCake) [reply utilities/pipelines/resourceRemoval/helper/Get-DeploymentTargetResourceList.ps1:111]: I'm not able to replicate the effect. I'd appreciate if we could have a chat next week to sort it out.

**Review by MrMCake — APPROVED** [review]

**Pull request merged**